libraries/standard/mqtt/src/mqtt.c

```diff
@@ -130,6 +234,447 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+    {
+        *pPublishState = newState;
+    }
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleKeepAlive( MQTTContext_t * const pContext )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint32_t now = 0U, keepAliveMs = 0U;
+
+    assert( pContext != NULL );
+    now = pContext->callbacks.getTime();
+    keepAliveMs = 1000U * ( uint32_t ) pContext->keepAliveIntervalSec;
+
+    /* If keep alive interval is 0, it is disabled. */
+    if( ( keepAliveMs != 0U ) && 
+        ( calculateElapsedTime( now, pContext->lastPacketTime ) > keepAliveMs ) )
+    {
+        if( pContext->waitingForPingResp )
+        {
+            /* Has time expired? */
+            if( calculateElapsedTime( now, pContext->pingReqSendTimeMs ) > 
+                pContext->pingRespTimeoutMs )
+            {
+                status = MQTTKeepAliveTimeout;
+            }
+        }
+        else
+        {
+            status = MQTT_Ping( pContext );
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleIncomingPublish( MQTTContext_t * const pContext,
+                                           MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t publishRecordState = MQTTStateNull;
+    uint16_t packetIdentifier;
+    MQTTPublishInfo_t publishInfo;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    status = MQTT_DeserializePublish( pIncomingPacket, &packetIdentifier, &publishInfo );
+    LogInfoWithArgs( "De-serialized incoming PUBLISH packet: DeserializerResult=%d", status );
+
+    if( status == MQTTSuccess )
+    {
+        publishRecordState = MQTT_UpdateStatePublish( pContext,
+                                                      packetIdentifier,
+                                                      MQTT_RECEIVE,
+                                                      publishInfo.qos );
+        LogInfoWithArgs( "State record updated. New state=%d.",
+                         publishRecordState );
+
+        /* Send PUBACK or PUBREC if necessary. */
+        status = sendPublishAcks( pContext,
+                                  packetIdentifier,
+                                  &publishRecordState );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* Provide publish info to application. */
+        pContext->callbacks.appCallback( pContext,
+                                         pIncomingPacket,
+                                         packetIdentifier,
+                                         &publishInfo );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleIncomingAck( MQTTContext_t * const pContext,
+                                       MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
```

Comment:
Would prefer this to be initialized to a failure value, if possible.

Reply:
Changed. The status is updated in each branch of the switch anyway so the initial value doesn't matter here, but I understand why it is a good practice.